ext/drx/scatter_gather_aarch64.c

```diff
@@ -38,6 +38,481 @@
+        DRX_CASE(stnt1d, 1, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(ld2b, 2, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld2h, 2, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld2w, 2, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld2d, 2, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(st2b, 2, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st2h, 2, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st2w, 2, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st2d, 2, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(ld3b, 3, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld3h, 3, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld3w, 3, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld3d, 3, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(st3b, 3, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st3h, 3, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st3w, 3, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st3d, 3, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(ld4b, 4, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld4h, 4, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld4w, 4, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld4d, 4, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(st4b, 4, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st4h, 4, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st4w, 4, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(st4d, 4, OPSZ_8, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(ld1rob, 1, OPSZ_1, false, DRX_NORMAL_FAULTING);
+
+        DRX_CASE(ld1rqb, 1, OPSZ_1, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld1rqh, 1, OPSZ_2, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld1rqw, 1, OPSZ_4, false, DRX_NORMAL_FAULTING);
+        DRX_CASE(ld1rqd, 1, OPSZ_8, false, DRX_NORMAL_FAULTING);
+#undef DRX_CASE
+
+    default: DR_ASSERT_MSG(false, "Invalid scatter/gather instruction");
+    }
+}
+
+/*
+ * Emit code to expand a scalar + vector gather load into a series of equivalent scalar
+ * loads.
+ * These instructions have memory operands of the form:
+ *     [<Xn|SP>, <Zm>.<Ts>{, <mod>}]
+ * where addresses to load/store each element are calculated by adding a base address
+ * from the scalar register Xn, to an offset read from the corresponding element of the
+ * vector index register Zm.
+ *
+ * if (is_load)
+ *     clear_inactive_elements(dst);
+ * for (e=first_active_element(); e < num_elements; e =  next_active_element()) {
+ *     if (is_load)
+ *         dst[e] = scalar_load(base, offsets[e], mod);
+ *     else
+ *         scalar_store(src[e], base, offsets[e], mod);
+ * }
+
+ * For example
+ *     ld1d   (%x0,%z26.d,lsl #3)[32byte] %p1/z -> %z27.d
+ * expands to:
+ *
+ *       dup    $0x00 lsl $0x00 -> %z27.d       ; Clear dst register
+ *       pfalse  -> %p0.b
+ *   loop:
```

Comment:
The trouble with unrolling this loop is that we don't know until run time how many iterations there are. The x86 version loops over each element of the vector but here we only loop over the active elements, and we don't know how many active elements there are at translation time because we don't know the value of the guarding predicate.

If we have access to the values of the predicate registers in the state restore function then we should be able to determine which element access faulted by looking at the value of the scratch_pred register.

Reply:
> The x86 version loops over each element of the vector but here we only loop over the active elements,

Well, if needed, we could loop over each element and skip the inactive elements on our own.

> If we have access to the values of the predicate registers in the state restore function then we should be able to determine which element access faulted by looking at the value of the scratch_pred register.

Hmm.. this should work. I like the loop better if it works.

I'm trying to recall if there was any other motivation for unrolling the loop in x86.

@derekbruening Looking at my notes from a couple years ago, I see
```
TODO: separate issue about making it a loop.
```
I don't see a filed issue though.

Maybe we discussed changing it to a loop but decided it's low priority?

Reply:
My notes only have links to the following issues:
- #2985: problems instrumenting AVX2 vpgatherdd
- #5005: Verify drx state restoration for scatter/gather expansion with other instrumentation
- #5036: Expand A64 SVE scatter/gather memory access instructions
- #5259: 32-bit support for scatter-gather variations that have qword index or value
I don't have any further info written down about a loop.

Reply:
Okay.

Since all iterations are in a loop, the instrumentation will be repeated too. So we need to mark the first output pc entry and subsequent ones differently (like we do for x86 repstr; see TRACE_TYPE_INSTR_MAYBE_FETCH). This was simpler on x86 because of the unrolling.

https://github.com/DynamoRIO/dynamorio/blob/1b697e5847b6f5b5c277a13aa1e5aeb0af44a97e/clients/drcachesim/tracer/instru.cpp#L79

Reply:
I assume drmemtrace specific changes will come later? Maybe add that to the PR description?

Reply:
We wanted to get rid of the MAYBE_FETCH concept altogether though: #4915, #4948.  So I'm not sure we want to add another instance that uses it if we can avoid it?  To avoid it we'd have to unroll the loop?

Reply:
We could unroll the loop or use one of the options discussed in https://github.com/DynamoRIO/dynamorio/issues/4915#issuecomment-1064242403 for repstr.

The simplest option seems to be D, which suggests removing the non-first entries in raw2trace.

Reply:
For now at least, I have unrolled the loop. If we have a different solution in the future we can revisit making it a loop if we want to.

Reply:
Thanks for making the change. I rather like the loop version so if we figure out a solution for repstr, we can consider using the same for scatter-gather (x86 and AArch64). Could you please also leave a comment on #4948 about this design decision?